charmbits/httpservice/server.go

```diff
@@ -0,0 +1,439 @@
+package httpservice
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/gocharm/charmbits/service"
+	"github.com/juju/gocharm/hook"
+)
+
+// server represents a running HTTP server.
+type server struct {
+	tomb tomb.Tomb
+
+	handlerInfo *handlerInfo
+	handler     Handler
+	stateDir    string
+
+	mu    sync.Mutex
+	state ServerState
+	// relationState holds the struct value with fields
+	// set by the setFieldValue argument to RegisterRelation
+	// calls.
+	relationState reflect.Value
+	httpListener  *handlerListener
+	httpsListener *handlerListener
+}
+
+// ServerState holds the state of a server - all the
+// parameters that might affect it.
+//
+// It is an implementation detail and only exposed
+// because net/rpc requires it.
+type ServerState struct {
+	HTTPPort  int
+	HTTPSPort int
+	CertPEM   string
+
+	// Arg holds the value passed to Service.Start, marshaled as JSON.
+	ArgData []byte
+
+	// RelationValues holds an entry for each relation
+	// defined in the second argument of the handler function
+	// passed to Register.
+	RelationValues map[string][]byte
+}
+
+func startServer(ctxt *service.Context, args []string, h *handlerInfo) (hook.Command, error) {
```

Comment:
Why not add this to the startServer0, I mean, combine them. It's hard to understand esp. as you return c, err in both cases. The code is actually redundant.